lib/coverband/collectors/view_tracker.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+require 'singleton'
+
+module Coverband
+  module Collectors
+    ###
+    # This class tracks view file usage via ActiveSupport::Notifications
+    #
+    # This is a port of Flatfoot, a project I open sourced years ago, but am now rolling into Coverband
+    # https://github.com/livingsocial/flatfoot
+    #
+    # TODO: test and ensure slim, haml, and other support
+    ###
+    class ViewTracker
+      DEFAULT_TARGET = Dir.glob('app/views/**/*.html.erb').reject { |file| file.match(/(_mailer)/) }
+      attr_accessor :target, :logged_views, :views_to_record
+      attr_reader :logger, :roots, :store, :ignore_patterns
+
+      def initialize(options = {})
+        raise NotImplementedError, 'View Tracker requires Rails 4 or greater' unless self.class.supported_version?
+
+        @project_directory = File.expand_path(Coverband.configuration.root)
+        @ignore_patterns = Coverband.configuration.ignore
+        @store = options.fetch(:store) { Coverband.configuration.store }
+        @logger = options.fetch(:logger) { Coverband.configuration.logger }
+        @target = options.fetch(:target) { DEFAULT_TARGET }
+
+        @roots = options.fetch(:roots) { Coverband.configuration.all_root_patterns }
+        @roots = @roots.split(',') if @roots.is_a?(String)
+
+        @logged_views = []
+        @views_to_record = []
+      end
+
+      def track_views(_name, _start, _finish, _id, payload)
+        if (file = payload[:identifier])
+          if newly_seen_file?(file)
+            logged_views << file
+            views_to_record << file if track_file?(file)
+          end
+        end
+
+        ###
+        # Annoyingly while you get full path for templates
+        # notifications only pass part of the path for layouts dropping any format info
+        # such as .html.erb or .js.erb
+        # http://edgeguides.rubyonrails.org/active_support_instrumentation.html#render_partial-action_view
+        ###
+        return unless (layout_file = payload[:layout])
+        return unless newly_seen_file?(layout_file)
+
+        logged_views << layout_file
+        views_to_record << layout_file if track_file?(layout_file)
+      end
+
```

Comment:
Does it make sense to extract a method here? Something like:

```ruby
def track_views(_name, _start, _finish, _id, payload)
  track_view(payload[:identifier])
  track_view(payload[:layout])
end

private

def track_view(file)
  if file && newly_seen_file?(file)
    logged_views << file
    views_to_record << file if track_file?(file)
  end
end
```

Not sure if that'll work.

Reply:
it does work, but this is the tight loop that runs during Rails web requests, multiple times, it runs on every render call. So we want to keep it as fast as possible. For that reason, I have avoided all extra method calls. I could document that perhaps?

Reply:
Interesting. Kind of surprised that a few method calls would make a difference. We can test this kind of thing out once we have some solid benchmarking.

Reply:
yeah, I haven't really benchmarked this, but I know when I built it a few years ago it went through a few perf iterations. I can work on getting more benchmarks... I do believe for a rails app with a partial in a loop say table row partials, this can end up with hundreds of calls in a single page render.